New features

luizoti opened the pull request.

luizoti committed:
update ignored

luizoti committed:
Update on the keys that characterize a keyboard

luizoti committed:
fix indexerror when device is removed

luizoti committed:
fix OSError when device is removed

luizoti committed:
Ability to remap based on the device name, in config

luizoti committed:
Merge branch 'master' of https://github.com/luizoti/xkeysnail

luizoti committed:
added missing args to on_key and
multipurpose_handler

luizoti committed:
added the transformed key to output

luizoti committed:
fix now print commant, sleep, buttons and combo

luizoti committed:
updates in launch function

luizoti committed:
Update setup.py

luizoti committed:
added except to and messege to missing config

luizoti committed:
now --config/-c is a required flag, added short flags

luizoti committed:
update in FileNotFoundError related with config.py

luizoti committed:
update variable names

luizoti committed:
Removed unused line

luizoti committed:
Merge branch 'master' of https://github.com/luizoti/xkeysnail

luizoti committed:
update

luizoti committed:
feat: kill other instancies by command

luizoti committed:
feat: auto search for config in predefined paths

luizoti committed:
feat: startup delay to help systemd service

luizoti committed:
update var names

luizoti committed:
pylint diable

luizoti committed:
update help messege

luizoti committed:
moved impoty sys to top

luizoti committed:
feat: another instace detection

luizoti committed:
update: detect if runing as sudo

luizoti committed:
update: config search will auto detect username

luizoti committed:
update: will exit if no --device or --watch is enable

luizoti committed:
update: changes in kill function

luizoti committed:
update: check moved to function

luizoti committed:
update: 'Device Added' removed from print_device_list

luizoti committed:
update: remove appdirs, add psutil

luizoti committed:
update: add more itens

luizoti committed:
update: function now search dir not file

luizoti committed:
update: use unix signal and lockfile to deal with other instancies

luizoti committed:
removed function runneruser

luizoti committed:
update: initial actions moved to __init__

luizoti committed:
Merge pull request #1 from luizoti/dev

update: initial actions moved to __init__

luizoti committed:
Update README.md

luizoti committed:
update: now xkeysnail will wait for config

luizoti committed:
Update README.md

luizoti committed:
Update README.md

luizoti committed:
update: wait_config does not have to wait all the time informed

luizoti committed:
update: better error capture

luizoti committed:
update: added pylint: disable

luizoti committed:
update: unused var converted to underscore

luizoti committed:
update: check another instance on startup

luizoti committed:
add: recomended .service file

luizoti committed:
fix: :bug: fix missing variable

This variables is imported from xkeysnail/info.py.

luizoti committed:
build: :heavy_plus_sign: add custon python-evdev repo install

luizoti committed:
refactor: :recycle: fully refactory all startup methods into a startupclass

luizoti committed:
fix: :bug: renamed external repo to fix pkg_resources.DistributionNotFound

Rename repository to follow package name aparently fix problems like: pkg_resources.DistributionNotFound: The 'python-evdev==1.4.1' distribution was not found and is required by xkeysnail

luizoti committed:
feat: :sparkles: primitive log module

this is a very primitive log module, this module can log as a decorator or simple messeges.

luizoti committed:
refactor: :art: refactored into a new class and placed in main module

luizoti committed:
style: :art: applyed pep8 rules and added docstrings, but not finished

luizoti committed:
style: :bookmark: Version 0.4.53

luizoti committed:
style: :bookmark: Version 0.4.53

luizoti committed:
Merge pull request #2 from luizoti/dev

Version 0.4.53

luizoti committed:
Merge branch 'main' of https://github.com/luizoti/xkeysnail into main

luizoti committed:
style: :fire: short messeges

luizoti committed:
style: :fire: short messeges

luizoti committed:
style: :art: splited imports

luizoti committed:
style: :art: splited imports

luizoti committed:
style: :art: added __name__ variable

luizoti committed:
style: :art: added __name__ variable

luizoti committed:
fix: :bug: dix default boot time

luizoti committed:
fix: :bug: dix default boot time

luizoti committed:
Version: 0.4.57

luizoti committed:
Version: 0.4.57

luizoti committed:
docs: :memo: update docstrings

luizoti committed:
fix: :bug: add xhost +SI:localuser:root to setup.py

luizoti committed:
fix: :bug: updates to fix problems in startup

luizoti committed:
shebang changed to env

luizoti committed:
refactor: :art: __author__ moved to info.py

luizoti committed:
docs: :memo: fix markdown violations

luizoti committed:
docs: :memo: update info.py

luizoti committed:
refactor: :fire: removed xlib modules

luizoti committed:
refactor: :recycle: move xlib methods to a standalone module

luizoti committed:
refactor: :recycle: autopep8

luizoti committed:
fix: :bug: fix print combo

luizoti committed:
refactor: :art: move functions to utils.py

luizoti committed:
fix: :bug: rename module name

luizoti committed:
refactor: :art: revise imports

luizoti committed:
refactor: :art: move all global variables to top

luizoti committed:
style: :recycle: autopep8

luizoti committed:
refactor: :fire: removed unnecessary variables

luizoti committed:
fix: :bug: fix UnboundLocalError

luizoti committed:
feat: :sparkles: messege for keys with no setup

luizoti committed:
fix: :bug: fix filemode to append

luizoti committed:
refactor: :art: update imports

luizoti committed:
fix: :bug: fix startup and lockpid creation

luizoti committed:
refactor: :sparkles: add logs to startp messages

luizoti committed:
fix: :bug: fix wmclass detection

Now in theory it is possible to extract wmclass, and wmname separately, more information like screen resolution and can be extracted as well.

luizoti committed:
Merge branch 'main' into dev

luizoti committed:
Merge pull request #3 from luizoti/dev

Version: 0.4.81

luizoti committed:
Bump Version: 0.4.81

luizoti committed:
Merge branch 'dev' of https://github.com/luizoti/xkeysnail into dev

luizoti committed:
Bump Version: 0.4.81

luizoti committed:
Update README.md

luizoti committed:
added more one exemple

luizoti committed:
update in error capture

luizoti committed:
added lower to wmname and wmclass

luizoti committed:
update syntax

luizoti committed:
change warning to info in log

luizoti committed:
update author

luizoti committed:
Merge pull request #4 from luizoti/dev

0.4.81

luizoti committed:
deleted info module

luizoti committed:
pylint updates

luizoti committed:
requirements moved to requirements.txt

luizoti committed:
Update instalation method

rbreaves commented:
So many edits.. why would anyone accept this PR? Maybe if you broke it up and actually said what each addition's purpose is. Not like I or anyone else other than mooz has the ability to merge PRs into this project as far as I know though.

luizoti commented:
> So many edits..

I hadn't noticed, amazing.

> why would anyone accept this PR?

I don't know and at the moment I'm not really caring about it.

> Maybe if you broke it up 

 I'm not in the mood, in fact, I uploaded this PR just in case someone wants to analyze it, no, if not, patience, my fork is working for me and that's fine.

> actually said what each addition's purpose is.

Read the title of each commit, it's an extremely succinct description, but I remember that I tried to make it as clear as I could.

> Not like I or anyone else other than mooz has the ability to merge PRs into this project as far as I know though.

I think the oldest PR in the repository is mine and it's been there for almost 1 years, the commits must be even older, I already know that mooz will most likely not follow through with the project, do what, xkeysnail is a good tool.

rbreaves commented:
Well things have seemed stagnate for awhile, but I am sure some of that is that Mooz does not want to break what's working for most people. I have broken master before forcing him to kick a commit of mine out, but has kept others.

I was not trying to be rude, it just seems like a lot for a single PR and perhaps the individual comments are enough for Mooz. I do not know. Might be good if mooz could find co-maintainers for this project though.

We do need to get Wayland and other things fixed I am sure.

Also I hadn't realized you had helped me with a PR before and I need to test out your changes, thank you for that.

luizoti commented:
Well, if he doesn't want to accept commits because he doesn't want to break main the project will die anyway, in that case it's better to continue with a fork. 

About Wayland, I think it won't be that hard to get it working, but the problem is, is Wayland working? Are there methods that deliver what we need to detect the wm_name, wm_class? I remember doing research a year or so ago and I didn't find much. I use KDE and I don't remember where the Wayland-supported version is.

All this needs to be studied. I think some xkeysnail modules need to be rewritten, the use of global makes everything very complex and confusing, I started, but I myself don't use all the functions and I don't know when they apply to be able to modify them and with that I gave up.a

rbreaves commented:
I think he's probably just not very active on github, but if you ping him over twitter he might get to it faster (as in measured in weeks or a couple of months still lol). What I did any ways, don't recall how I mentioned it exactly, but it probably helped that my PR's were relatively small.

And yea, Wayland is a moving target, but I did start a thread w/ some preliminary work in it. If someone doesn't address then I will try and circle back to it.

luizoti commented:
Maybe I'll try calling him and see what happens, then I'll see that.

Your proposal was made on commig https://github.com/rbreaves/xkeysnail/commit/2a8dd0e54f53751b0daaf8370c4a9538942b48d2, right?

Maybe it's possible to implement with https://pypi.org/project/pywayland/ if the api has methods that deliver the wm_class and etc...

rbreaves commented:
Yea, that commit was my proposal, but I think it may no longer be valid as I do not think that API is still be exposed in that way. I think an extension or some type of wrapper will need to be created now, which sucks as I use Budgie and cannot use Gnome extensions.. but I think there is still some method to make it work.

luizoti commented:
Dude, wrap this command in a function, and put it inside the xkeysnail, after that, just call it inside the get_active_window_wm_class function.

```gdbus call -e -d org.gnome.Shell -o /org/gnome/Shell -m org.gnome.Shell.Eval global.get_window_actors\(\)[`gdbus call -e -d org.gnome.Shell -o /org/gnome/Shell -m org.gnome.Shell.Eval global.get_window_actors\(\).findIndex\(a\=\>a.meta_window.has_focus\(\)===true\) | cut -d"'" -f 2`].get_meta_window\(\).get_wm_class\(\) | cut -d'"' -f 2 > /tmp/active_win```

Hence, it looks for a way to detect if it is running on X11 or Wayland to decide when to run this function or the logic of X11.

If it works, it will be lighter and more practical without needing to sleep or read files.

luizoti commented:
In the future, if an api for wayland appears that delivers everything that is needed, just implement the whole thing using it.

rbreaves commented:
What I was trying to say though is that the active window is not supposed to be readable directly via "org.gnome.Shell" going forward, if I recall the devs are closing that off or already have to gnome extension access instead. So I think you'd need a helper app outside of that gdbus call.

luizoti commented:
True, I had forgotten about this behavior of dbus not working as root, it really wouldn't be possible to integrate directly.

joshgoebel commented:
@luizoti Did you ever speak with the current maintainer?  I'm trying to figure out if this project is still actively maintained or not... looking at the last commit and state of PRs/issues it looks like the original author hasn't been involved in quite some time, or am I missing something?

luizoti commented:
Diretamente não tentei ainda, ao menos não lembro ter tentado, para mim o dev original não deve voltar a tocar no projeto.

> @luizoti Did you ever speak with the current maintainer? I'm trying to figure out if this project is still actively maintained or not... looking at the last commit and state of PRs/issues it looks like the original author hasn't been involved in quite some time, or am I missing something?

I haven't tried it directly yet, at least I don't remember trying, for me the original dev shouldn't touch the project again.

joshgoebel commented:
I just reached out to the author via email and asked if they would be interested in passing on maintainer-ship of the project... if so, would you be interested in helping out?

luizoti commented:
If he passes you, I can help when there's time, no problem.

joshgoebel commented on xkeysnail/input.py:
```diff
@@ -17,12 +17,24 @@ def get_devices_list():
+                Key.ENTER,
+                Key.KEY_0,
+                Key.KEY_5,
+                Key.KEY_9,
```
Is there a reason A and Z were removed?

luizoti replied:
If I'm not mistaken it was because it adds support for unconventional keyboards, which don't have A and Z like the PS3 BD remote.

Initially in my setup all keyboards and the PS3 BD remote started to be detected.

joshgoebel commented:
@luizoti You interested in a fork of the project if the author is simply non-responsive?  I looked at the first 10 - 20 of your commits here and most of this stuff I would merge immediately without question (I even had some of the exact same thoughts as I reviewed the code).  Some of the new feature stuff might require a bit more discussion.  I'm guessing if I reviewed the entire list that trend would continue - ie most of this code would be mergeable as-is.

I don't need any more projects that I'm the sole maintainer of but if we got a small team together I think we could perhaps push this forward.

My first priority would be a test suite so we can make sure we avoid regressions moving forward in how the engine works.  After there is a solid test suite in place then refactoring the core becomes a lot less scary.

luizoti commented:
Some commits related to configurations, I would probably drop today, simply to reduce code, another I have to see what was done.

A few things I find interesting to look at in a refactoring:

1. Removing the need for global variables, sometimes they make everything complex and confusing.
2. Extend remap functionalities so that it is possible to remap each device according to wm_name, wm_class, if they are in full screen and if necessary to have different profiles for each situation.

These are ideas, possibilities.

joshgoebel commented:
I see you adding version numbers in some of your commits, have you actually released your version anywhere in public where such things would matter?

luizoti commented:
I just changed the version on github.

luizoti committed:
project renamed to keyscrew

joshgoebel commented on xkeysnail/__init__.py:
```diff
@@ -1,5 +1,22 @@
+
+    killed = None
+    try:
+        for proc in psutil.process_iter():
```
We need to figure our what our "best practices" runtime story is... if someone is supposed to just run us from systemd then I'm not sure any of this process management stuff is necessary - systemd is more than capable of starting/stopping/signaling processes all on it's own - since we don't fork.

If someone is running us (as themselves) from say `.xinitrc` (I think that could be another way to go) then this is one line of code to do this in bash, I'm not sure we need to build the checking into the keymapper itself, etc.  I'm trying to imagine when one would have lots of extra processes running around that they had to deal with.   Sometimes I wish your commit messages were more than one liners - and explained the thinking and rational for each change. :-)

What was your use case for adding this?

luizoti replied:
I did this exactly because using systemd it couldn't kill the process when necessary.
Another case is that the user needs it for the process that was running in the background, for whatever reason.

If there is another solution for systemd to handle this better, this can be removed.

joshgoebel replied:
Doesn't systemd send correct signals when you ask it to stop a service?  Could it be that we just need to handle those signals on the Python side?  I'm unsure.

luizoti replied:
I opened the code yesterday in this snippet because I didn't remember for sure what it was, this snippet is from the -k command line, it has nothing to do with systemd.

The system signals are already registered elsewhere in the program.

joshgoebel commented on xkeysnail/__init__.py:
```diff
@@ -113,7 +114,7 @@ def cli_main():
 
     # Make sure that user have root privilege
-    if not has_access_to_uinput():
+    if not has_access_to_uinput() or os.getuid() != 0:
```
I don't follow this... surely access to uinput is ALL that matters here, what is the extra check doing?  We might have access without being root - it's east to configure that.

luizoti replied:
At the time, the goal was probably to make it more verbose and clear why it couldn't be executed, it's been that way for some time.

As it doesn't affect the remap loop, I don't think it's critical in terms of performance.

```    @staticmethod
    def check_if_access_to_uinput():
        """Staticmethod to check if _uinput is accessible."""
        if _UINPUT:
            return True
        else:
            log_msg("Failed to open `uinput` in write mode.")
            if os.getuid() != 0:
                log_msg("Make sure that you have executed keyscrew with root privileges.")
                log_msg("Such as: sudo keyscrew -c config.py")
            sys.exit(1)
```

joshgoebel replied:
Much of my code review (and hence the questions) is about clarity, not just performance. :)  I want to 100% understand every line of code.

joshgoebel commented on xkeysnail/xlib_info.py:
```diff
@@ -2,60 +2,53 @@
+            time =- 1
+        except Xlib.error.ConnectionClosedError as Xlib_ConnectionClosedError:
+            log_msg('Xlib: %s ' % Xlib_ConnectionClosedError)
+            return Display()
```
What does this do? If Display just blew up on line 32, why would it suddenly be working now on line 37?  Wouldn't we want to instead fall thru to line 38 and return None?

luizoti replied:
That's what should happen, I probably kept that line there by mistake.

luizoti committed:
Update xlib_info.py

luizoti committed:
removed useless exit

luizoti committed:
reorder imports

luizoti committed:
remove space from staticmethod

luizoti committed:
remove useless variable

joshgoebel commented:
Don't feel you have to change every little thing I point out... I'm probably going to start with 0.4.0 as-is, pull over as much of your work as seems reasonable and then build on top of that...  A lot of times I'm just asking to understand what's going on - not expecting you to immediately go and make a fix.

Are you running xkeysnail in systemd but AS your own user? I'm a bit confused.

joshgoebel commented:
```
+                if check_device.name == "py-evdev-uinput":
+                    # Exclude evdev device, we use for output emulation,
```

How did this get added, was this a problem you were actually having, you accidentally included the output device?

luizoti commented:
> Don't feel you have to change every little thing I point out... I'm probably going to start with 0.4.0 as-is, pull over as much of your work as seems reasonable and then build on top of that... A lot of times I'm just asking to understand what's going on - not expecting you to immediately go and make a fix.
> 
> Are you running xkeysnail in systemd but AS your own user? I'm a bit confused.

O evdev cria esse dispositivo e precisa ser ignorado em certos momentos.

luizoti commented:
> ```
> +                if check_device.name == "py-evdev-uinput":
> +                    # Exclude evdev device, we use for output emulation,
> ```
> 
> How did this get added, was this a problem you were actually having, you accidentally included the output device?

The evdev creates this device and needs to be ignored at certain times.

luizoti commented:
> Don't feel you have to change every little thing I point out... I'm probably going to start with 0.4.0 as-is, pull over as much of your work as seems reasonable and then build on top of that... A lot of times I'm just asking to understand what's going on - not expecting you to immediately go and make a fix.
> 
> Are you running xkeysnail in systemd but AS your own user? I'm a bit confused.

As root, systemd was configured to give the user dbus access, to generate desktop notifications.

luizoti committed:
refactor: :zap: Methods related to xlib, aggregated into a single class.

The plan is to reduce Display calls, which used to happen every time you searched for wm_class and wm_name, now the connection is made at the beginning and only returns information when requested.  This change also allows to return the resolution and current status of the window in relation to being in fullscreen or not.

luizoti committed:
Merge branch 'main' of https://github.com/luizoti/xkeysnail into main

luizoti committed:
removed print
